yugabyte/src/yugabyte/ysql/client.clj

```diff
@@ -0,0 +1,205 @@
+(ns yugabyte.ysql.client
+  "Helper functions for working with YSQL clients."
+  (:require [clojure.java.jdbc :as j]
+            [clojure.pprint :refer [pprint]]
+            [clojure.string :as str]
+            [clojure.tools.logging :refer [info warn]]
+            [jepsen.util :as util]
+            [jepsen.control.net :as cn]
+            [jepsen.reconnect :as rc]
+            [dom-top.core :as dt]
+            [wall.hack :as wh]
+            [slingshot.slingshot :refer [try+ throw+]]))
+
+(def timeout-delay "Default timeout for operations in ms" 20000)
+(def max-timeout "Longest timeout, in ms" 40000)
+
+(def isolation-level "Default isolation level for txns" :serializable)
+
+(def ysql-port 5433)
+
+(defn db-spec
+  "Assemble a JDBC connection specification for a given Jepsen node."
+  [node]
+  {:dbtype         "postgresql"
+   :dbname         "postgres"
+   :classname      "org.postgresql.Driver"
+   :host           (name node)
+   :port           ysql-port
+   :user           "postgres"
+   :password       ""
+   :loginTimeout   (/ max-timeout 1000)
+   :connectTimeout (/ max-timeout 1000)
+   :socketTimeout  (/ max-timeout 1000)})
+
+(defn close-conn
+  "Given a JDBC connection, closes it and returns the underlying spec."
+  [conn]
+  (when-let [c (j/db-find-connection conn)]
+    (.close c))
+  (dissoc conn :connection))
+
+(defn conn-wrapper
+  "Constructs a network client for a node, and opens it"
+  [node]
+  (rc/open!
+    (rc/wrapper
+      {:name  node
+       :open  (fn open []
+                (util/timeout max-timeout
+                              (throw (RuntimeException.
+                                       (str "Connection to " node " timed out")))
+                              (util/retry 0.1
+                                          (let [spec  (db-spec node)
+                                                conn  (j/get-connection spec)
+                                                spec' (j/add-connection spec conn)]
+                                            (assert spec')
+                                            spec'))))
+       :close close-conn
+       :log?  true})))
+
+(defn exception->op
```

Comment:
Looks like this is what `with-errors` macro does for YCQL. Can we rename this as `with-errors` as well and make it a macro to reuse the common pattern for both APIs ?